src/bin/dual-iir.rs

```diff
@@ -2,19 +2,18 @@
 #![no_std]
 #![no_main]
 
-use stabilizer::{hardware, net};
-
-use miniconf::Miniconf;
 use serde::Deserialize;
 
 use dsp::iir;
-use hardware::{
-    Adc0Input, Adc1Input, AdcCode, AfeGain, Dac0Output, Dac1Output, DacCode,
-    DigitalInput0, DigitalInput1, InputPin, SystemTimer, AFE0, AFE1,
+use stabilizer::{
+    hardware::{
+        hal, setup, Adc0Input, Adc1Input, AdcCode, AfeGain, Dac0Output,
+        Dac1Output, DacCode, DigitalInput0, DigitalInput1, InputPin,
+        SystemTimer, AFE0, AFE1,
+    },
+    net::{Miniconf, NetworkUsers, Telemetry, TelemetryBuffer, UpdateState},
```

Comment:
In general, I'm more of a fan of directly importing things where used from the direct source. It makes it easier to trace where things are coming from.

If you export `Miniconf` in `stabilizer::net`, then someone has to look at `stabilizer::net` first to see that `Miniconf` comes from `miniconf::Miniconf` and isn't some alias of it.

However, when we import `miniconf::Miniconf` directly where used, it's quick and easy to reference exactly where Miniconf is being derived from.

In general, my preference is to import things from the source wherever possible to alleviate the mental burden of going through the manual lookup

Reply:
I'm ok with keeping the namespace hierarchy and not `pub use` re-exporting.
But that isn't the style you have been following so far at all. Currently it's a happy mixture of `pub use` and `pub mod`.
It would remove the wildcard `pub use` re-exports but makes the actual use in the application quite a bit harder to read.

Reply:
I don't think I'm completely following your logic here. This is purely in regard to keeping crate-dependency-level `use` statements where-used.

Reply:
There are two topics of namespace management here. One is the re-exporting of to remove namespace hierarchy levels. E.g. `TelemetryBuffer`. This is `pub mod` to export the module vs `pub use` to re-export the item.
The other is re-exporting interfaces/traits from other crates. If we don't do that, it becomes extremely annoying to copy over and match exact dependencies in e.g. `fls`. Having to do so is pointless because the correct dependency is right there in `stabilizer`. In the case of macros (Miniconf, Serialize, Deserialize) it unfortunately doesn't work very well but for traits it's very convenient to be able to get them re-exported from `stabilizer` and very inconvenient to have to cross-depend on a crate and manage+track that dependency. Re-exporting also seems to be somewhat common, especially if the downstream crate will generally only need the trait because and through the intermediate crate.

Reply:
Given that we're treating stabilizer as a library, I can get on board with this choice. Thanks for taking the time to explain your reasoning here.

Reply:
Context:
[serde-json-core](https://github.com/rust-embedded-community/serde-json-core/blob/986dcab749210bd1a4b25f6e75b1e5e549428fa4/src/lib.rs#L73) re-exports `heapless`. Maybe `miniconf` should re-export `serde-json-core` as well.